tephra-core/src/main/java/org/apache/tephra/runtime/TransactionDistributedModule.java

```diff
@@ -41,14 +41,15 @@
 
   @Override
   protected void configure() {
+    // some of these classes need to be non-singleton in order to create a new instance during leader() in
+    // TransactionService
     bind(SnapshotCodecProvider.class).in(Singleton.class);
-    bind(TransactionStateStorage.class).annotatedWith(Names.named("persist"))
-      .to(HDFSTransactionStateStorage.class).in(Singleton.class);
-    bind(TransactionStateStorage.class).toProvider(TransactionStateStorageProvider.class).in(Singleton.class);
+    bind(TransactionStateStorage.class).annotatedWith(Names.named("persist")).to(HDFSTransactionStateStorage.class);
+    bind(TransactionStateStorage.class).toProvider(TransactionStateStorageProvider.class);
 
-    bind(TransactionManager.class).in(Singleton.class);
-    bind(TransactionSystemClient.class).to(TransactionServiceClient.class).in(Singleton.class);
```

Comment:
Since `TransactionServiceClient` can contain pool of thrift clients, it is better to have it as a singleton.